src/git/repo.go

```diff
@@ -0,0 +1,590 @@
+package git
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/git-town/git-town/src/command"
+	"github.com/git-town/git-town/src/util"
+	"github.com/git-town/git-town/test/helpers"
+)
+
+var repoInCurrentDir *Repo
+
+// RepoInCurrentDir provides a Repo instance in the current working directory.
+func RepoInCurrentDir() *Repo {
+	if repoInCurrentDir == nil {
+		repoInCurrentDir = &Repo{
+			Dir:   ".",
+			Shell: &command.ShellInCurrentDir{},
+		}
+	}
+	return repoInCurrentDir
+}
+
+// Repo is a Git repository that exists inside a Git environment.
+type Repo struct {
+
+	// Dir contains the path of the directory that this repository is in.
+	Dir string
+
+	// Shell runs console commands in this repo.
+	command.Shell
+
+	// configCache contains the Git Town configuration to use.
+	// This value is lazy loaded. Please use Configuration() to access it.
+	configCache *Configuration
+
+	// currentBranch contains the current Git branch we are on in this repo
+	currentBranch string
+
+	// list of the remote branches
+	remoteBranches []string
+
+	// indicates whether the remoteBranches property has been initialized
+	remoteBranchesInitialized bool
+}
+
+// AddRemote adds the given Git remote to this repository.
+func (repo *Repo) AddRemote(name, value string) error {
```

Comment:
How did you choose what functions are in this file? I'd prefer we start with only what are using in source as I think a number of these are used only for tests

Reply:
After trying out a few alternatives, I have settled on moving as many as possible Git operations to the production side, especially the "normal" ones that don't introduce dependencies on test frameworks. Ideally, all Git operations are in the same place. This makes maintenance so much easier. I don't want to have to look in two places to see if a certain Git functionality exists. I especially don't want external contributors to have to do this. Simpler is better here.

This `AddRemote` function is a good example. It is probably only used in tests at this point. But if we move it to the `test` package, and later start needing it in production, we have to move this method over to the production codebase again. That's unnecessary churn. Even worse, every time we change something, we have to check if any Git function should be moved around now. This means an ongoing need for labor-intensive maintenance.

The only downside I can see here is that this increases the binary size. We are talking changes in the order of having a 12.7 MB binary vs a 12.6 MB binary here. Such unused code pieces areeasily be removed by tree shaking. And if binary size is a concern, that's a separate optimization effort that should be data driven. I don't like the idea that binary size dictates how we organize our source code.  🙂

Reply:
I agree with the argument of trying to avoid having to look in two places and having to move code between places based on test / production usage.

I do think it can be confusing to find test only code though in a production folder (given the general separation pattern seen on other projects). I'm willing to try this out but think we need a comment at the top of GitRepo explaining this contains some repository functions only used in tests

Reply:
The way I look at this is that this new `git` package is a Go SDK for command-line Git. It allows running Git commands using the Git version that is installed in the system. It provides really high-level abstractions like "add a remote named foo", rather than letting users run `"git remote add foo"` themselves. This package doesn't care whether it is called from production or test code. It is so generic that it could as well be an external package if we wanted to extract it. 

This new `git` package should not introduce dependencies on other packages. That's why I separate code that depends on `godog` out into the `test` package. Because only our `test` package should depend on `godog`.